manta-trusted-setup/src/pairing.rs

```diff
@@ -0,0 +1,105 @@
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Pairing
+
+use ark_ec::{AffineCurve, PairingEngine};
+use ark_ff::PrimeField;
+use core::iter;
+
+/// Pairing Configuration
+pub trait Pairing {
+    /// Underlying Scalar Field
+    type Scalar: PrimeField;
+
+    /// First Group of the Pairing
+    type G1: AffineCurve<ScalarField = Self::Scalar> + Into<Self::G1Prepared>;
+
+    /// First Group Pairing-Prepared Point
+    type G1Prepared;
+
+    /// Second Group of the Pairing
+    type G2: AffineCurve<ScalarField = Self::Scalar> + Into<Self::G2Prepared>;
+
+    /// Second Group Pairing-Prepared Point
+    type G2Prepared;
+
+    /// Pairing Engine Type
+    type Pairing: PairingEngine<
+        G1Affine = Self::G1,
+        G2Affine = Self::G2,
+        G1Prepared = Self::G1Prepared,
+        G2Prepared = Self::G2Prepared,
+    >;
+
+    /// Returns the base G1 generator for this configuration.
+    fn g1_prime_subgroup_generator() -> Self::G1;
+
+    /// Returns the base G2 generator for this configuration.
+    fn g2_prime_subgroup_generator() -> Self::G2;
+}
+
+/// Pair from a [`PairingEngine`]
+type Pair<P> = (
+    <P as PairingEngine>::G1Prepared,
+    <P as PairingEngine>::G2Prepared,
+);
+
+/// Pairing Engine Extension
+pub trait PairingEngineExt: PairingEngine {
+    /// Evaluates the pairing function on `pair`.
+    #[inline]
+    fn eval(pair: &Pair<Self>) -> Self::Fqk {
+        Self::product_of_pairings(iter::once(pair))
+    }
+
+    /// Checks if `lhs` and `rhs` evaluate to the same point under the pairing function.
+    #[inline]
+    fn has_same(lhs: &Pair<Self>, rhs: &Pair<Self>) -> bool {
+        Self::eval(lhs) == Self::eval(rhs)
+    }
+
+    /// Checks if `lhs` and `rhs` evaluate to the same point under the pairing function, returning
+    /// `Some` with prepared points if the pairing outcome is the same. This function checks if
+    /// there exists an `r` such that `(r * lhs.0 == rhs.0) && (lhs.1 == r * rhs.1)`.
+    #[inline]
+    fn same<L1, L2, R1, R2>(lhs: (L1, L2), rhs: (R1, R2)) -> Option<(Pair<Self>, Pair<Self>)>
+    where
+        L1: Into<Self::G1Prepared>,
+        L2: Into<Self::G2Prepared>,
+        R1: Into<Self::G1Prepared>,
+        R2: Into<Self::G2Prepared>,
+    {
+        let lhs = (lhs.0.into(), lhs.1.into());
+        let rhs = (rhs.0.into(), rhs.1.into());
+        Self::has_same(&lhs, &rhs).then_some((lhs, rhs))
+    }
+
+    /// Checks if the ratio of `(lhs.0, lhs.1)` from `G1` is the same as the ratio of
+    /// `(lhs.0, lhs.1)` from `G2`.
+    #[inline]
+    fn same_ratio<L1, L2, R1, R2>(lhs: (L1, R1), rhs: (L2, R2)) -> bool
+    where
+        L1: Into<Self::G1Prepared>,
+        L2: Into<Self::G2Prepared>,
+        R1: Into<Self::G1Prepared>,
+        R2: Into<Self::G2Prepared>,
+    {
+        Self::has_same(&(lhs.0.into(), rhs.1.into()), &(lhs.1.into(), rhs.0.into()))
+    }
+}
+
+impl<E> PairingEngineExt for E where E: PairingEngine {}
```

Comment:
```suggestion
impl<E> PairingEngineExt for E where E: PairingEngine {}

#[cfg(test)]
mod test {
    use ark_bls12_381::{Bls12_381, Fr, G1Affine, G2Affine};
    use ark_ec::{AffineCurve, ProjectiveCurve};

    use crate::pairing::PairingEngineExt;

    /// Sanity check on the behavior of the pairing check
    #[test]
    fn same_ratio_test() {
        let g1_base = G1Affine::prime_subgroup_generator();
        let g2_base = G2Affine::prime_subgroup_generator();
        let scalar = Fr::from(4u8);
        let g1_scaled = g1_base.mul(scalar).into_affine();
        let g2_scaled = g2_base.mul(scalar).into_affine();
        let lhs = (g1_base, g2_scaled);
        let rhs = (g1_scaled, g2_base);

        assert!(Bls12_381::same(lhs, rhs).is_some())
    }
}
```
I find it hard to keep track of our convention for how the arguments should be passed into `same`, so I think it's nice to have a test that makes it clear to see.